src/main/java/com/google/devtools/build/lib/analysis/RuleContext.java

```diff
@@ -1390,19 +1411,18 @@ public boolean targetPlatformHasConstraint(ConstraintValueInfo constraintValue)
     } else {
       return parseExecProperties(
           execProperties,
-          toolchainContexts == null ? ImmutableSet.of() : toolchainContexts.getExecGroups());
+          toolchainContexts == null ? null : toolchainContexts.getExecGroups());
     }
   }
 
   /**
    * Parse raw exec properties attribute value into a map of exec group names to their properties.
    * The raw map can have keys of two forms: (1) 'property' and (2) 'exec_group_name.property'. The
-   * former get parsed into the target's default exec group, the latter get parsed into their
-   * relevant exec groups.
+   * former get parsed into the default exec group, the latter get parsed into their relevant exec
+   * groups.
    */
-  private static ImmutableMap<String, ImmutableMap<String, String>> parseExecProperties(
-      Map<String, String> rawExecProperties, Set<String> execGroups)
-      throws InvalidExecGroupException {
+  private static Map<String, Map<String, String>> parseExecGroups(
```

Comment:
Wow, I hate Map<?, Map<?, ?>> so much, but I see this predates your change so leave it. I will probably send a followup shortly to change this to a [Table](https://github.com/google/guava/wiki/NewCollectionTypesExplained#table).

Reply:
Acknowledged. I didn't know about Table (never wrote Java at Google). I would be happy to fix this - please let me know if you'd like me to do this as part of this PR - but you probably have more context.